cli/lint_reporter.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+//! This module defines an abstraction LintReporter, which is something that knows
+//! how to report observed lints (for example, pretty print or json).
+use crate::colors;
+use crate::fmt_errors;
+use crate::lint_serialize::SerializableLint;
+use deno_core::ErrBox;
+use deno_lint::diagnostic::LintDiagnostic;
+use serde::Serialize;
+
+pub enum LintReporterKind {
+  Console,
+
+  // `sort` is used to enforce determinism in tests.
+  Json { sort: bool },
+}
+
+pub trait LintReporter {
+  fn visit(&mut self, d: &LintDiagnostic);
+  fn visit_error(&mut self, file_path: &str, err: &ErrBox);
+  fn close(&mut self);
+}
+
+pub fn create_reporter(kind: LintReporterKind) -> Box<dyn LintReporter + Send> {
+  match kind {
+    LintReporterKind::Console => Box::new(ConsoleLintReporter::new()),
+    LintReporterKind::Json { sort } => Box::new(JsonLintReporter::new(sort)),
+  }
+}
+
+#[derive(Serialize)]
+struct LintError {
+  file_path: String,
+  message: String,
+}
+
+struct ConsoleLintReporter {
+  lint_count: u32,
+}
+
+impl ConsoleLintReporter {
+  fn new() -> ConsoleLintReporter {
+    ConsoleLintReporter { lint_count: 0 }
+  }
+}
+
+impl LintReporter for ConsoleLintReporter {
+  fn visit(&mut self, d: &LintDiagnostic) {
+    self.lint_count += 1;
+
+    let pretty_message =
+      format!("({}) {}", colors::gray(&d.code), d.message.clone());
+
+    let message = fmt_errors::format_stack(
+      true,
+      &pretty_message,
+      Some(&d.line_src),
+      Some(d.location.col as i64),
+      Some((d.location.col + d.snippet_length) as i64),
+      &[fmt_errors::format_location(
+        &d.location.filename,
+        d.location.line as i64,
+        d.location.col as i64,
+      )],
+      0,
+    );
+
+    eprintln!("{}\n", message);
+  }
+
+  fn visit_error(&mut self, file_path: &str, err: &ErrBox) {
+    eprintln!("Error linting: {}", file_path);
+    eprintln!("   {}", err);
+  }
+
+  fn close(&mut self) {
+    match self.lint_count {
+      1 => eprintln!("Found 1 problem"),
+      n if n > 1 => eprintln!("Found {} problems", self.lint_count),
+      _ => (),
+    }
+  }
+}
+
+#[derive(Serialize)]
+struct JsonLintReporter {
+  lints: Vec<SerializableLint>,
```

Comment:
I think we should use `diagnostics` rather than `lints` because that is what these are called in deno_lint and TypeScript.

So:
- `lints` -> `diagnostics`
- `SerializableLint` -> `SerializableDiagnostic`

Reply:
Agreed on `lints` -> `diagnostics`, but I feel that `SerializableLint` is superfluous... `deno_lint::Diagnostic` is already JSON-serializable, so it should be only a matter of stringifying it